# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/nsBaseWidget.cpp
# Commit: 641765261f49
# Full Hash: 641765261f499a1a4712c0c02a14b58b69ecabaa
# Author: sotaro <sotaro.ikeda.g@gmail.com>
# Date: 2019-07-26 21:36:38
# Regressor Bug: 1569110
# File Overlap Count: 1
# Description:
#   Bug 1569110 - Add ClientLayerManager::Initialize() r=nical
#   
#   Some initialization handlings of ClientLayerManager exists in nsBaseWidget::CreateCompositor(). It is not good. Move them to ClientLayerManager::Initialize().
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D39476
# ==============================================================================

diff -r a6fa09658817 -r 641765261f49 widget/nsBaseWidget.cpp
--- a/widget/nsBaseWidget.cpp	Fri Jul 26 13:25:35 2019 +0000
+++ b/widget/nsBaseWidget.cpp	Fri Jul 26 09:31:57 2019 +0000
@@ -1307,6 +1307,16 @@
         gfx::GPUProcessManager::Get()->DisableWebRender(
             wr::WebRenderError::INITIALIZE);
       }
+    } else if (lm->AsClientLayerManager() && mCompositorSession) {
+      bool shouldAccelerate = ComputeShouldAccelerate();
+      TextureFactoryIdentifier textureFactoryIdentifier;
+      lm->AsClientLayerManager()->Initialize(
+          mCompositorSession->GetCompositorBridgeChild(), shouldAccelerate,
+          &textureFactoryIdentifier);
+      if (textureFactoryIdentifier.mParentBackend ==
+          LayersBackend::LAYERS_NONE) {
+        DestroyCompositor();
+      }
     }
 
     // We need to retry in a loop because the act of failing to create the
@@ -1374,39 +1384,9 @@
                LayersBackend::LAYERS_WR);
     ImageBridgeChild::IdentifyCompositorTextureHost(textureFactoryIdentifier);
     gfx::VRManagerChild::IdentifyTextureHost(textureFactoryIdentifier);
-  }
-
-  ShadowLayerForwarder* lf = lm->AsShadowForwarder();
-  if (lf) {
-    // lf is non-null if we are creating a ClientLayerManager above
-    TextureFactoryIdentifier textureFactoryIdentifier;
-    PLayerTransactionChild* shadowManager = nullptr;
-
-    nsTArray<LayersBackend> backendHints;
-    gfxPlatform::GetPlatform()->GetCompositorBackends(ComputeShouldAccelerate(),
-                                                      backendHints);
-
-    bool success = false;
-    if (!backendHints.IsEmpty()) {
-      shadowManager = mCompositorBridgeChild->SendPLayerTransactionConstructor(
-          backendHints, LayersId{0});
-      if (shadowManager->SendGetTextureFactoryIdentifier(
-              &textureFactoryIdentifier) &&
-          textureFactoryIdentifier.mParentBackend !=
-              LayersBackend::LAYERS_NONE) {
-        success = true;
-      }
-    }
-
-    if (!success) {
-      NS_WARNING("Failed to create an OMT compositor.");
-      DestroyCompositor();
-      mLayerManager = nullptr;
-      return;
-    }
-
-    lf->SetShadowManager(shadowManager);
-    lm->UpdateTextureFactoryIdentifier(textureFactoryIdentifier);
+  } else if (lm->AsClientLayerManager()) {
+    TextureFactoryIdentifier textureFactoryIdentifier =
+        lm->GetTextureFactoryIdentifier();
     // Some popup or transparent widgets may use a different backend than the
     // compositors used with ImageBridge and VR (and more generally web
     // content).
